src/ObjCRuntime/Runtime.cs

```diff
@@ -1735,11 +1737,23 @@ internal static IntPtr GetProtocolForType (Type type)
 			throw new ArgumentException ($"'{type.FullName}' is an unknown protocol");
 		}
 
-		[BindingImpl (BindingImplOptions.Optimizable)]
 		internal static bool IsUserType (IntPtr self)
 		{
 			var cls = Class.object_getClass (self);
+			if (!usertype_cache.TryGetValue (cls, out var result)) {
```

Comment:
This would need a lock to protect against race conditions.

Reply:
Both the key and the value are value types, and I believe ConditionalWeakTable only works with reference types.

Also the key represents the native Class object, and that will never be freed, so I don't see much value in trying to keep track of lifetimes on the managed side.